packages/stardog-graphql-language-server/src/GraphQlLanguageServer.ts

```diff
@@ -0,0 +1,243 @@
+        };
+      } else {
+        textEditRange = {
+          start: document.positionAt(tokenAtCursor.startOffset),
+          end: document.positionAt(tokenAtCursor.endOffset + 1),
+        };
+      }
+
+      return TextEdit.replace(textEditRange, replacement);
+    };
+
+    // Completions are collected in this way (pushing, etc.) for
+    // speed/complexity reasons (fewer map/filter/flatten operations needed).
+    const allCompletions = [];
+    candidates.forEach((candidate) => {
+      const completionItem = this.getCompletionItem(
+        candidate,
+        replaceTokenAtCursor
+      );
+      if (!completionItem) {
+        return;
+      }
+
+      if (Array.isArray(completionItem)) {
+        allCompletions.push(...completionItem.filter(Boolean));
+      } else {
+        allCompletions.push(completionItem);
+      }
+    });
+
+    return uniqBy(allCompletions, 'label');
+  }
+
+  private getCompletionItem(
+    candidate: CompletionCandidate,
+    tokenReplacer: (
+      replacement: string,
+      replacementRange?: CompletionCandidate['replacementRange']
+    ) => TextEdit
+  ): CompletionItem | CompletionItem[] | void {
+    const { PATTERN } = candidate.nextTokenType;
+
+    if (typeof PATTERN !== 'string') {
+      // This token uses either a custom pattern-matching function or a regex
+      // and we therefore cannot use its pattern for autocompletion.
+      return;
+    }
+
+    return {
+      label: PATTERN,
+      kind: CompletionItemKind.EnumMember,
+      textEdit: tokenReplacer(PATTERN, candidate.replacementRange),
+    };
+  }
+
+  onContentChange(
+    { document }: TextDocumentChangeEvent,
+    parseResult: ReturnType<
+      AbstractLanguageServer<
+        StardogGraphQlParser | StandardGraphQlParser
+      >['parseDocument']
+    >
+  ) {
+    const { uri } = document;
+    const content = document.getText();
+
+    if (!content.length) {
+      this.connection.sendDiagnostics({
+        uri,
+        diagnostics: [],
+      });
+      return;
+    }
+
+    const { errors } = parseResult;
+    const diagnostics = this.getParseDiagnostics(document, errors);
+
+    this.connection.sendDiagnostics({
+      uri,
+      diagnostics,
+    });
+  }
+
+  // OVERRIDE for special EOF error token handling in case of SPARQL errors
+  getParseDiagnostics(document: TextDocument, errors: ISemanticError[]) {
+    const content = document.getText();
+
+    return errors.map(
+      (error): Diagnostic => {
+        const { message, context, token, name } = error;
+        const ruleStack = context ? context.ruleStack : null;
+        const source =
+          ruleStack && ruleStack.length > 0
+            ? ruleStack[ruleStack.length - 1]
+            : null;
+        const constructedDiagnostic: Partial<Diagnostic> = {
+          message,
+          source,
+          severity: DiagnosticSeverity.Error,
+        };
+
+        // For the GraphQL parser, internal SPARQL errors can have EOF tokens,
```

Comment:
why are there internal SPARQL errors in the graphql parser?

Reply:
Because Stardog's GraphQL extensions allow embedding SPARQL. Our parser is able to check that you've entered valid SPARQL there (or not).